### PR TITLE
request info: improve downstream/upstream timings

### PR DESCRIPTION
**Pull request by rodaine**
This patch continues on the work in #743, adding the various downstream/upstream timings described in the [gRPC access log payload](https://github.com/envoyproxy/data-plane-api/blob/4b93d3697e788dd2810613289ef83788dfe7da44/envoy/config/filter/accesslog/v2/accesslog.proto#L50-L86) to `RequestInfo`. As a result, the existing timings (`requestReceived`, `responseReceived`, and `duration`) have been removed in favor of these new ones. The access log (both gRPC and file-based) utilities have also been updated to use and/or emit the new values. The duration are now exposed as `MonotonicTime`, allowing a consumer to choose the appropriate cast (ie, access log formatter uses ms, while gRPC timestamps support nanosecond precision); a utility class was added to facilitate the conversion.

Couple questions that came up while working on this:
* Does the `RequestInfo` on the TcpProxy need to have timings populated? Does that make sense?
* Should the `UpstreamRequest`'s `RequestInfo` have timings attached as well? Does that make sense?
* For `firstUpstreamRxByteReceived`, it looks like the closest I could get it to the actual first byte is inside `UpstreamRequest`'s `decodeHeaders`. Is there a better, more accurate place to capture the time?

*Risk Level*: Low-Medium (changes how some existing timings are calculated)
*Testing*: Unit
*Docs Changes*: N/A (these are still marked unimplemented in the data plane, but if desired I can add documentation for these timings)
*Release Notes*: N/A (unless it's worthwhile to mention that RESPONSE_DURATION and DURATION are calculated differently in the existing logs?)

/cc @derekargueta (#2654)
Signed-off-by: Chris Roche <croche@lyft.com>

### CODE REVIEW

File: ci/build_setup.sh [hunk 1]
```diff
@@ -6,7 +6,7 @@ set -e
 
 export PPROF_PATH=/thirdparty_build/bin/pprof
 
-[ -z "${NUM_CPUS}" ] && NUM_CPUS=`grep -c ^processor /proc/cpuinfo`
+[ -z "${NUM_CPUS}" ] && NUM_CPUS=`nproc`
```
Comment (ggreenway) [inline ci/build_setup.sh]: This should be in a separate PR.

File: source/common/access_log/access_log_formatter.cc [hunk 2]
```diff
@@ -150,23 +161,13 @@ RequestInfoFormatter::RequestInfoFormatter(const std::string& field_name) {
...
     field_extractor_ = [](const RequestInfo::RequestInfo& request_info) {
-      Optional<std::chrono::microseconds> duration = request_info.responseReceivedDuration();
-      if (duration.valid()) {
-        return std::to_string(
-            std::chrono::duration_cast<std::chrono::milliseconds>(duration.value()).count());
-      } else {
-        return UnspecifiedValueString;
-      }
+      return AccessLogFormatUtils::durationToString(request_info,
+                                                    request_info.firstUpstreamRxByteReceived());
```
Comment (ggreenway) [inline source/common/access_log/access_log_formatter.cc]: For RESPONSE_DURATION, shouldn't this be looking at when the last Upstream byte was received?

**Comment by ggreenway** [comment]
Also, TSAN build is failing

File: source/common/access_log/grpc_access_log_impl.cc [hunk 3]
```diff
@@ -162,22 +159,74 @@ void HttpGrpcAccessLog::log(const Http::HeaderMap* request_headers,
...
               request_info.startTime().time_since_epoch())
               .count()));
-  if (request_info.requestReceivedDuration().valid()) {
+
+  if (request_info.lastDownstreamRxByteReceived().valid()) {
     common_properties->mutable_time_to_last_rx_byte()->MergeFrom(
-        Protobuf::util::TimeUtil::MicrosecondsToDuration(
-            request_info.requestReceivedDuration().value().count()));
+        Protobuf::util::TimeUtil::NanosecondsToDuration(
+            RequestInfo::TimingUtils::duration<std::chrono::nanoseconds>(
```
Comment (ggreenway) [inline source/common/access_log/grpc_access_log_impl.cc]: Please add a helper function for the common part:
```
void helper(some_type& dest, const Optional<MonotonicTime>& src)
   if (src.valid()) {
       dst.MergeFrom(
          Protobuf::util::TimeUtil::NanosecondsToDuration(
             RequestInfo::TimingUtils::duration<std::chrono::nanoseconds>(
                 request_info, src)
                 .value()
                 .count()));
```

Or similar

File: source/common/access_log/access_log_formatter.cc (hunk 2 above)
Comment (rodaine) [reply source/common/access_log/access_log_formatter.cc]: RESPONSE_DURATION was measured off of the old `responseReceived` value, which was captured in [`onUpstreamHeaders`](https://github.com/envoyproxy/envoy/pull/2660/files#diff-a00f27eaffcd28ed6b74102912eb1d65L629), so it measured TTFB, not the complete response. Unless I'm misinterpreting this...

File: source/common/request_info/request_info_impl.h [hunk 4]
```diff
@@ -15,39 +17,80 @@ struct RequestInfoImpl : public RequestInfo {
...
   RequestInfoImpl(Http::Protocol protocol) : RequestInfoImpl() { protocol_ = protocol; }
 
-  // RequestInfo::RequestInfo
-  SystemTime startTime() const override { return start_time_; }
+  const SystemTime& startTime() const override { return start_time_; }
+
+  const MonotonicTime& startTimeMonotonic() const override { return start_time_monotonic_; }
+
+  const Optional<MonotonicTime>& lastDownstreamRxByteReceived() const override {
+    return last_rx_byte_received_;
```
Comment (ggreenway) [inline source/common/request_info/request_info_impl.h]: Please rename all the members to match their corresponding function names, for clarity.  Eg last_downstream_rx_byte_received_ here.

File: ci/build_setup.sh (hunk 1 above)
Comment (rodaine) [reply ci/build_setup.sh]: Whoops! Slipped in. Fixed.

File: source/common/request_info/request_info_impl.h [hunk 5]
```diff
@@ -15,39 +17,80 @@ struct RequestInfoImpl : public RequestInfo {
...
+
+  const Optional<MonotonicTime>& lastDownstreamTxByteSent() const override {
+    return last_downstream_tx_byte_sent_;
   }
 
+  void lastDownstreamTxByteSent(MonotonicTime time) override {
+    last_downstream_tx_byte_sent_.value(time);
+  }
+
+  void finalize(MonotonicTime time) override { end_time_.value(time); }
```
Comment (ggreenway) [inline source/common/request_info/request_info_impl.h]: Should this be renamed to finalTimeMonotonic (and in interface, of course) to match the pattern in the rest of the interface?

File: source/common/router/router.cc [hunk 6]
```diff
@@ -416,13 +416,9 @@ void Filter::maybeDoShadowing() {
 void Filter::onRequestComplete() {
   downstream_end_stream_ = true;
   downstream_request_complete_time_ = std::chrono::steady_clock::now();
```
Comment (ggreenway) [inline source/common/router/router.cc]: Can this be computed from one of the existing time values on the request info?

File: include/envoy/request_info/request_info.h [hunk 7]
```diff
@@ -71,63 +71,123 @@ class RequestInfo {
...
-  virtual SystemTime startTime() const PURE;
+  virtual const SystemTime& startTime() const PURE;
 
   /**
-   * @return duration from request start to when the entire request was received from the
-   * downstream client in microseconds.
+   * @return the monotonic time that the first byte of the request was received. Duration
+  calculations should be made relative to this value.
+  */
+  virtual const MonotonicTime& startTimeMonotonic() const PURE;
```
[the remark this answers is not in the document]
Comment (rodaine) [reply include/envoy/request_info/request_info.h]: Wanted to allow the caller to choose the desired duration units when performing the cast. But the same could be achieved by returning nanoseconds and allowing a caller to cast to a lower resolution unit. It would simplify things (and not require the helper utility).

File: source/common/request_info/request_info_impl.h (hunk 5 above)
Comment (rodaine) [reply source/common/request_info/request_info_impl.h]: Was leaving it generic if there's other finalization tasks to be added in the future. YAGNI, I suppose; will rename.

File: source/common/router/router.cc (hunk 6 above)
Comment (rodaine) [reply source/common/router/router.cc]: Totally, though if we convert the return values to durations, we won't have access to the time points.

Comment (rodaine) [reply source/common/router/router.cc]: Oh I take that back, we immediately calculate a duration from it. LastDownstreamTxByteSent should be good.

**Comment by rodaine** [comment]
> It would be great if you could add all the new fields to text-based access logs as well, to keep parity between the two.

Was going to do this in a follow up PR for #2654, but I can do it here if desired.

File: include/envoy/request_info/request_info.h [hunk 8]
```diff
@@ -71,63 +71,123 @@ class RequestInfo {
   /**
    * @return the time that the first byte of the request was received.
    */
-  virtual SystemTime startTime() const PURE;
+  virtual const SystemTime& startTime() const PURE;
```
Comment (alyssawilk) [inline include/envoy/request_info/request_info.h]: If we leave this as is, maybe change this to firstDownstreamRxByteReceived for consistency?

If it's returning an absolute time where everything else does duration you can leave it "special" :-)

File: include/envoy/request_info/request_info.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (alyssawilk) [reply include/envoy/request_info/request_info.h]: Yeah, I think if you're up for doing that it would address I wonder if the switch from SystemTime to MonotonicType addresses @mattklein123 's TODO for cleaning up time units.

File: source/common/filter/tcp_proxy.cc [hunk 9]
```diff
@@ -120,6 +120,8 @@ TcpProxy::TcpProxy(TcpProxyConfigSharedPtr config, Upstream::ClusterManager& clu
       upstream_callbacks_(new UpstreamCallbacks(this)) {}
 
 TcpProxy::~TcpProxy() {
+  request_info_.finalize(std::chrono::steady_clock::now());
```
Comment (alyssawilk) [inline source/common/filter/tcp_proxy.cc]: Hmmm, this is interesting.  I was going to ask that you make sure all the new fields are tracked for TCP proxy session too, but I see that we're already not consistently doing that.    @ggreenway given the logical things (first byte read, last byte sent) make sense for TCP proxy, but the request/response naming does not, do you think  we should do neutral naming (nix request/response) and track it here, neutral naming and track things for TCP proxy some day, or just comment that it's HTTP only?

File: source/common/http/conn_manager_impl.cc [hunk 10]
```diff
@@ -941,6 +944,7 @@ void ConnectionManagerImpl::ActiveStream::encodeHeaders(ActiveStreamEncoderFilte
 #endif
 
   // Now actually encode via the codec.
+  request_info_.firstDownstreamTxByteSent(std::chrono::steady_clock::now());
```
Comment (alyssawilk) [inline source/common/http/conn_manager_impl.cc]: Can we document at the API level that our times generated when bytes are serialized to the appropriate codec?  This is almost certainly a good enough approximation for 99.99% of Envoy users and as long as we've docced up "this won't let you know when your control frames are backed up due to flow control" I think we're good.

File: include/envoy/request_info/request_info.h [hunk 11]
```diff
@@ -71,63 +71,123 @@ class RequestInfo {
...
-   * @return the microseconds duration of the first byte received to the last byte sent.
+   * @return the monotonic clock time when the first byte of the response is received from upstream.
    */
-  virtual std::chrono::microseconds duration() const PURE;
+  virtual const Optional<MonotonicTime>& firstUpstreamRxByteReceived() const PURE;
+
+  /**
+   * Set the monotonic clock time when the first byte of the response is received from upstream.
+   */
+  virtual void firstUpstreamRxByteReceived(MonotonicTime) PURE;
```
Comment (alyssawilk) [inline include/envoy/request_info/request_info.h]: For all the upstream cases, we should clarify what we do in case of retries, where we've sent headers to the backend, it does an early failure, and we ship things off a second time.  I assume second pass gets recorded, but it's worth doccing and testing.

File: include/envoy/request_info/request_info.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (rodaine) [reply include/envoy/request_info/request_info.h]: Yeah, Matt and I discussed this approach initially, but after implementing it felt really awkward to use (having to calculate the duration yourself). I'm cool with internalizing it into RequestInfo.

File: source/common/filter/tcp_proxy.cc (hunk 9 above)
Comment (ggreenway) [reply source/common/filter/tcp_proxy.cc]: I think all the names in the new interfaces are in terms of upstream/downstream bytes, which should be fine.  The interface comments mention request/response, so that could be cleaned up.

I think in the tcpproxy, we should think about what makes sense.  For instance, firstDownstreamRx and firstUpstreamTx should always be the same time, so does it make sense to track both?  And for last byte in either direction, we don't know when it happens until the connection closes, so we'd have to take a timestamp on every single rx/tx, and use the last one after the connection closes.  

I think my preference is to sort out all those details in a separate PR.  This one maintains parity in tcp_proxy with previous state, which I this is good enough for this PR.

File: include/envoy/request_info/request_info.h [hunk 12]
```diff
@@ -71,63 +71,123 @@ class RequestInfo {
...
+
+  /**
+   * Set the monotonic clock time when the first byte of the response is sent downstream.
+   */
+  virtual void firstDownstreamTxByteSent(MonotonicTime) PURE;
+
+  /**
+   * @return the monotonic clock time when the last byte of the response is sent downstream.
+   */
+  virtual const Optional<MonotonicTime>& lastDownstreamTxByteSent() const PURE;
```
[the remark this answers is not in the document]
Comment (ggreenway) [reply include/envoy/request_info/request_info.h]: Most of the times in request_info have been optionals since I added access logging to tcp_proxy, because they don't make sense in all contexts, so this lets the interface return "no value set" in those cases.

**Comment by ggreenway** [comment]
@rodaine I'm fine with doing the text-logging in a followup PR.

File: include/envoy/request_info/request_info.h (hunk 8 above)
Comment (rodaine) [reply include/envoy/request_info/request_info.h]: I'll leave it "special" and convert all the other methods to return nanos instead. I think that gives us the most bang for our buck.

File: source/common/access_log/grpc_access_log_impl.cc (hunk 3 above)
Comment (rodaine) [reply source/common/access_log/grpc_access_log_impl.cc]: by calling `mutable_*`, it actually creates an empty message for that field, so this particular form doesn't work. With the switch to the RequestInfo returning nanos, however, it's pretty DRY'd up though.

File: include/envoy/request_info/request_info.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (rodaine) [reply include/envoy/request_info/request_info.h]: Also, from discussion with @mattklein123, if we don't actually make an upstream request (eg, an inline static response), we don't want upstream-specific timings.

File: source/common/router/router.cc (hunk 6 above)
Comment (rodaine) [reply source/common/router/router.cc]: Leaving these in for now. They're a bit tangled in the tests and will benefit from being done in a separate PR, if that's fine?

File: source/common/http/conn_manager_impl.cc (hunk 10 above)
Comment (rodaine) [reply source/common/http/conn_manager_impl.cc]: Do you mean in data-plane-api? I can create a PR there to mention the retry behavior, and I think the [flow control](https://github.com/envoyproxy/data-plane-api/blob/master/envoy/config/filter/accesslog/v2/accesslog.proto#L84) comments already exist.

**Comment by rodaine** [comment]
@ggreenway @alyssawilk I think this is ready for another look!

File: source/common/router/router.cc (hunk 6 above)
Comment (ggreenway) [reply source/common/router/router.cc]: Yes, I think that's fine

File: include/envoy/request_info/request_info.h [hunk 13]
```diff
@@ -68,66 +68,125 @@ class RequestInfo {
    */
   virtual void onUpstreamHostSelected(Upstream::HostDescriptionConstSharedPtr host) PURE;
 
+  /**
+   * @return the # of body bytes received in the request.
```
Comment (alyssawilk) [inline include/envoy/request_info/request_info.h]: super nitty: would prefer number be spelled out

File: include/envoy/request_info/request_info.h [hunk 14]
```diff
@@ -68,66 +68,125 @@ class RequestInfo {
...
+   */
+  virtual uint64_t bytesReceived() const PURE;
+
+  /**
+   * @return the protocol of the request.
+   */
+  virtual const Optional<Http::Protocol>& protocol() const PURE;
+
+  /**
+   * Set the request's protocol.
```
Comment (alyssawilk) [inline include/envoy/request_info/request_info.h]: We tend towards individual argument comments so
@param protocol ...
here and for other setter parameters below.

File: include/envoy/request_info/request_info.h [hunk 15]
```diff
@@ -68,66 +68,125 @@ class RequestInfo {
...
+   */
+  virtual void protocol(Http::Protocol protocol) PURE;
```
Comment (alyssawilk) [inline include/envoy/request_info/request_info.h]: I thought we mostly name things like this setProtocol but grepping for Set it looks like we're about 50/50.  I think setFoo is more clear than foo but given we're not consistent please consider this an optional suggestion :-)

File: include/envoy/request_info/request_info.h [hunk 16]
```diff
@@ -68,66 +68,125 @@ class RequestInfo {
...
   /**
-   * @return duration from request start to when the entire request was received from the
-   * downstream client in microseconds.
+   * @return the monotonic time that the first byte of the request was received. Duration
+  calculations should be made relative to this value.
+  */
+  virtual MonotonicTime startTimeMonotonic() const PURE;
+
+  /**
+   * @return the monotonic clock time when the last byte of the request was received.
```
Comment (alyssawilk) [inline include/envoy/request_info/request_info.h]: I see a bunch of comments about monotonic clock time with nanoseconds args.  Stale comments?

File: source/common/access_log/access_log_impl.cc [hunk 17]
```diff
@@ -81,8 +83,15 @@ bool StatusCodeFilter::evaluate(const RequestInfo::RequestInfo& info, const Http
 }
 
 bool DurationFilter::evaluate(const RequestInfo::RequestInfo& info, const Http::HeaderMap&) {
+  Optional<std::chrono::nanoseconds> final = info.finalTimeMonotonic();
+
+  if (!final.valid()) {
```
Comment (alyssawilk) [inline source/common/access_log/access_log_impl.cc]: I definitely think we should do error handling here but should we consider not finalizing valid or a bug?  It's especially weird that if you don't finalize, every time you evaluate you might get a different result.

File: source/common/http/conn_manager_impl.cc (hunk 10 above)
Comment (alyssawilk) [reply source/common/http/conn_manager_impl.cc]: Request redacted - I'd missed the existing comments and as you say they're pretty good :-)

File: include/envoy/request_info/request_info.h (hunk 13 above)
Comment (rodaine) [reply include/envoy/request_info/request_info.h]: Yep, that was an existing comment just mangled in the diff. Will fix for both received and sent.

File: include/envoy/request_info/request_info.h (hunk 15 above)
Comment (rodaine) [reply include/envoy/request_info/request_info.h]: The existing methods on RequestInfo were overloaded, so I was being consistent (at least locally).

File: include/envoy/request_info/request_info.h (hunk 16 above)
Comment (rodaine) [reply include/envoy/request_info/request_info.h]: Yep, will update

File: source/common/access_log/access_log_impl.cc (hunk 17 above)
Comment (rodaine) [reply source/common/access_log/access_log_impl.cc]: So in an earlier commit I rolled back (before PR) I had assertions to make sure finalTimeMonotonic wasn't called until it was set (and wasn't an optional value) but that seems like a weird way for this code to fail. I could throw instead? Not sure what makes the most sense.

File: include/envoy/request_info/request_info.h (hunk 15 above)
Comment (alyssawilk) [reply include/envoy/request_info/request_info.h]: Consistency is the best argument :-)  Let's leave it as-is then.

File: source/common/access_log/access_log_impl.cc (hunk 17 above)
Comment (mattklein123) [reply source/common/access_log/access_log_impl.cc]: I would consider it a bug and just do `ASSERT`. It's a programming error that we should be catching in tests.

File: include/envoy/request_info/request_info.h [hunk 18]
```diff
@@ -68,66 +68,134 @@ class RequestInfo {
...
 
+  /**
+   * @return the number of body bytes received in the request.
+   */
+  virtual uint64_t bytesReceived() const PURE;
+
+  /**
+   * @return the protocol of the request.
+   */
+  virtual const Optional<Http::Protocol>& protocol() const PURE;
```
Comment (mattklein123) [inline include/envoy/request_info/request_info.h]: As long as you are moving a lot of stuff around in here, can you switch this to not return by reference? It will be faster. Same for `responseCode()` and anything else that is an optional of a small primitive type.

File: source/common/request_info/request_info_impl.h [hunk 19]
```diff
@@ -15,39 +17,91 @@ struct RequestInfoImpl : public RequestInfo {
...
   RequestInfoImpl(Http::Protocol protocol) : RequestInfoImpl() { protocol_ = protocol; }
 
-  // RequestInfo::RequestInfo
   SystemTime startTime() const override { return start_time_; }
 
-  const Optional<std::chrono::microseconds>& requestReceivedDuration() const override {
-    return request_received_duration_;
+  MonotonicTime startTimeMonotonic() const override { return start_time_monotonic_; }
+
+  Optional<std::chrono::nanoseconds> duration(const Optional<MonotonicTime>& time) const {
```
Comment (mattklein123) [inline source/common/request_info/request_info_impl.h]: pass without reference

File: source/common/router/router.cc [hunk 20]
```diff
@@ -818,21 +812,33 @@ void Filter::UpstreamRequest::decode100ContinueHeaders(Http::HeaderMapPtr&& head
 }
 
 void Filter::UpstreamRequest::decodeHeaders(Http::HeaderMapPtr&& headers, bool end_stream) {
+  // TODO(rodaine): This is actually measuring after the headers are parsed and not the first byte.
+  parent_.callbacks_->requestInfo().firstUpstreamRxByteReceived(std::chrono::steady_clock::now());
```
Comment (mattklein123) [inline source/common/router/router.cc]: Do we ever do anything other than pass `std::chrono::steady_clock::now()` to any of these functions? I wonder ic it would better to actually have request info snap the current time and make these functions more like `onFirstUpstreamRxByteReceived()`. The benefit of that also is you can ASSERT it is never called twice to populate the optional twice. I think this would be an improvement so throwing it out there. (Obviously this applies to all cases of this pattern).

File: include/envoy/request_info/request_info.h (hunk 14 above)
Comment (alyssawilk) [reply include/envoy/request_info/request_info.h]: Also not sure if this one was missed or if you're still working, so bumping it just in case :-)

Comment (rodaine) [reply include/envoy/request_info/request_info.h]: Missed it (the diff really stirred up all of these)! Just going to fix it for all methods here and be done with it 😅

File: source/common/router/router.cc (hunk 20 above)
Comment (rodaine) [reply source/common/router/router.cc]: Yeah, other than tests it's exclusively `steady_clock::now()`. I'm cool with this change. As for the ASSERTs though, in the cases of retries, we do expect it to be called multiple times, no?

Comment (mattklein123) [reply source/common/router/router.cc]: When I originally commented on this I thought about retries, which is kind of an interesting case. I actually don't think the asserts should be an issue for retries. Here is why:

1) In the upstream log case, we should be getting discrete data points for each try.
2) In the downstream log case, the router should only ever being to forward (and time) a single request. If this is not currently the case, you should be able to rejigger the router code slightly to make it so.

I think adding ASSERTs is a good thing to do here. We can always remove them later if need be. You have been looking at this code a lot more recently so feel free to correct me if what I'm saying isn't true.

File: include/envoy/request_info/request_info.h [hunk 21]
```diff
@@ -175,21 +175,26 @@ class RequestInfo {
...
-   * @param time the monotonic clock time when the last byte of the response is sent downstream.
+   * Sets the time when the last byte of the response is sent downstream.
    */
-  virtual void lastDownstreamTxByteSent(MonotonicTime time) PURE;
+  virtual void onLastDownstreamTxByteSent() PURE;
 
   /**
-   * @param time the end time for the request. This method is called once the request has been fully
+   * @return the total duration of the request (i.e., when the request's ActiveStream is destroyed)
+   * and may be longer than lastDownstreamTxByteSent.
```
Comment (alyssawilk) [inline include/envoy/request_info/request_info.h]: and may -> which may

File: source/common/router/router.cc [hunk 22]
```diff
@@ -818,21 +813,33 @@ void Filter::UpstreamRequest::decode100ContinueHeaders(Http::HeaderMapPtr&& head
 }
 
 void Filter::UpstreamRequest::decodeHeaders(Http::HeaderMapPtr&& headers, bool end_stream) {
+  // TODO(rodaine): This is actually measuring after the headers are parsed and not the first byte.
```
Comment (alyssawilk) [inline source/common/router/router.cc]: When you fix this for 100-continue etc. please fix the sending-to-downstream side as well.
